### PR TITLE
Fixes magfest/panels#103: Adds custom jinja environment that lets plugins override templates

### PR DESCRIPTION
**Pull request by RobRuana**
~This may be the wrong approach to solving [this issue](https://github.com/magfest/panels/issues/103). I'd like to hear everyone's thoughts about this approach. Here's what I did:~

1. ~Copied `registration/form.html` to `registration/baseform.html`~
2. ~Replaced the contents of `registration/form.html` with `{% extends "registration/baseform.html" %}`~
3. ~Added some `{% block %}` extension points that other plugins can use to insert fields~

See https://github.com/magfest/panels/pull/121 for details on how this works for other plugins.

See [this comment](https://github.com/magfest/ubersystem/pull/2884#issuecomment-334875911) for updated details.

### CODE REVIEW

**Comment by kitsuta** [comment]
This adds another layer to our already immensely complex registration form setup.

Like, now we have:
* base.html
* preregbase.html, which extends base.html
* preregistration/form.html, which extends preregbase.html and includes regform.html, which itself includes regextra.html
* registration/form.html, which extends baseform.html, which extends base.html; this form also includes regform.html, which itself includes regextra.html
* a few other pages that extend either base or base-admin (which is just base with admin_area set to true) and include regform.html

So loading registration/form.html involves loading literally five separate templates. We already had that with our preregistration/form.html, but if possible I'd really like to avoid adding Yet Another Layer onto all this.

Also, no matter what we do, we can only add to a template _once_, because of the way sideboard overwrites templates. So this method means that the panel plugin would only add to the registration form if no other plugin tried to do so. Again, we already deal with this kind of conflict for regextra.html, we don't need yet another version of that.

How about this:
1. Leave everything as-is. We can use regextra.html and check if admin_area is true to show things only on the admin form.
2. As always, regextra.html should only go in 'event' plugins due to the "only one template will be loaded" situation. You can technically include it in feature plugins, but this may cause Issues if sideboard loads plugins out of order.
3. In the panels plugin, create a new template. Call it something like panel_applications_list.html. Use that admin_area check described in step 1 so that the contents of the template are only loaded on admin pages.
4. In the event plugins, _include that new template_ in regextra.html.
5. Now you have the new stuff inserted with minimal duplication and maximum separation -- each event will have to include the panels template into regextra.html separately, which is a pain, but I think that's the smallest downside we can get away with right now.

**Comment by binary1230** [comment]
Minor note, potentially unrelated and maybe not something we can do right this second, but regform.html needs to die, and a chunk of the jinja refactor was done specifically to allow that to happen.  There's a github issue detailing how to do it which would also help simplify some situations like this one

**Comment by RobRuana** [comment]
Don't have energy for a full response right now, but I totally agree with @kitsuta. This is the wrong approach, for all the reasons outlined above. Not sure how to approach this. I have some ideas, gonna do some investigating tomorrow. 

Will report back!

😄

**Comment by RobRuana** [comment]
I've come up with a solution that I think works really well for our plugin system. It requires minimal changes, and is totally transparent to developers writing plugin templates.

# Limitations in the Current System

1. If multiple plugins have templates with the same relative path, only the first template found by the jinja environment will be loaded
   * If both `uber` and `panels` have a template at `registration/form.html`, then only the `panels` template will be loaded
2. Templates are distinguished solely by their relative path
   * Every time the jinja environment tries to load `registration/form.html` it will always load the one from `panels` – this prevents the `panels` `registration/form.html` from extending the `uber` `registration/form.html`

We have historically worked around the above limitations by:
1. Creating a `base_form.html` that contains the bulk of the logic, which each plugin may extend
2. Creating a `form_extra.html` that contains additional logic, which each plugin may override

This has worked to an extent, but the above work-arounds still suffer from the central limitations of the system. Only one plugin at a time may extend `base_form.html` or override `form_extra.html`

# Template Inheritance Requirements

1. Any number of plugins should be able to extend a template in a parent plugin
   * _templates in `guests` should be able to extend templates in `panels` and `uber`_
   * _templates in `panels` should be able to extend templates in `uber`_
2. Any number of plugins should be able to update or override a `{% block %}`
   * _Both `guests` and `panels` should be able to append fields to `registration/form.html` in `uber`_
3. Avoiding `base_*.html` templates is preferable (_like `registration/base_form.html`_)
4. Avoiding `*_extra.html` templates is preferable (_like `registration/form_extra.html`_)

# Proposal for Improving Template Inheritance

We can remove the limitations in the current system by creating custom jinja `Environment` and `Loader` classes

```python
# uber/jinja.py

class PluginEnvironment(jinja2.Environment):
    pass

class PluginLoader(jinja2.FileSystemLoader):
    pass

class JinjaEnv:
    @classmethod
    def _init_env(cls):
        env = PluginEnvironment(
            autoescape=True,
            loader=PluginLoader(cls._template_dirs))
```

The purpose of these classes is to further distinguish templates by not only their relative path, _but also the plugin from which they are loaded_. So when we call `environment.load_template('registration/form.html')`, the environment knows which plugin is requesting the template, and can load the template from the right place.

Assume there is a `registration/form.html` template in `guests`, `panels`, `hotel`, and `uber`...

The following sequence of events describes how it works:
1. The server receives a GET request for `/uber/registration/form`
2. `PluginEnviroment.load_template('registration/form.html')` is called
3. `PluginEnviroment` queries the current call stack to determine if the call was made from another template (via `{% include %} or {% extends %}`)
4. Since the call was NOT made from another template (it came from the GET request) the standard template loader is used, and `hotel/templates/registration/form.html` is returned
5. The template at `hotel/templates/registration/form.html` calls `{% extend 'registration/form.html' %}`
6. Once again the `PluginEnviroment` queries the current call stack to determine if the call was made from another template
7. Since the call WAS made from another template, the `PluginLoader` skips trying to load that template again, and continues searching up the chain of template search paths
8. The `PluginLoader` next finds the template at `guests/templates/registration/form.html` which itself calls `{% extend 'registration/form.html' %}`
9. The process is repeated until finally `uber/templates/registration/form.html` is loaded

## Example Usage

#### Hotel Plugin
`plugins/hotel/hotel/templates/registration/form.html`
```
{% extends "registration/form.html" %}
{% block content -%}
  <h1>HOTEL!</h1>
  {{ super() }}
{%- endblock %}
```

#### Panels Plugin
`plugins/panels/panels/templates/registration/form.html`
```
{% extends "registration/form.html" %}
{% block content -%}
  <h1>PANELS!</h1>
  {{ super() }}
{%- endblock %}
```

#### Guests Plugin
`plugins/guests/guests/templates/registration/form.html`
```
{% extends "registration/form.html" %}
{% block content -%}
  <h1>GUESTS!</h1>
  {{ super() }}
{%- endblock %}
```

#### Uber Plugin
`plugins/uber/uber/templates/registration/form.html`
```
{% extends "base.html" %}{% set admin_area=True %}
{% block title %}Attendee Form - {{ attendee.full_name }}{% endblock %}
{% block content %}
<!--
  Big huge template, unchanged from the way it is now...
-->
{% endblock %}
```

#### Rendered Result
screen shot 2017-10-06 at 4 46 11 pm

# Conclusion

I have working code that does what I described above! That screenshot was rendered for-real using the new system 😄 

We should be able to get rid of a lot of our `base_*.html` and `*_extra.html` wonkiness!

**Comment by kitsuta** [comment]
Whoa. That's rad.

One question -- is there going to be any shenanigans with load order? In other words, will the "HOTEL!", "GUESTS!", and "PANELS!" in this example be displayed in a different order every time you refresh, or every time you restart the server? This is totally something we can work around so I'm 110% for this solution, we'll just want to be aware of it if it's going to be a thing.

**Comment by binary1230** [comment]
Plugin load order should be deterministic even for unprioritized plugins, because we sort the list of unprioritized plugins on startup.  Only situation things could change is when we add or remove or rename plugins, which is not often.

The priority_plugins ini sideboard setting should probably be used here just to be super safe.

Otherwise, yea server restarts and such shouldn't change the load order of plugins

**Comment by RobRuana** [comment]
I've pushed the code that actually makes this work. In answer to the two questions:
1. The load order of plugins is deterministic, so it will never change, even between server reloads
2. There should be no appreciable performance impact from the proposed system (though we should probably do some additional profiling to confirm that)

File: uber/jinja.py [hunk 1]
```diff
@@ -1,4 +1,106 @@
...
+            if filename in requesting_templates:
+                # If the file is already in the call stack, ignore it
+                continue
+
+            f = open_if_exists(filename)
+            if f is None:
+                continue
+            try:
+                contents = f.read().decode(self.encoding)
+            finally:
```
Comment (kitsuta) [inline uber/jinja.py]: Should there be an `except` block here? What happens if there's a problem loading a template? Is that passed up the stack correctly so we see the exception rendered?

File: uber/jinja.py [hunk 2]
```diff
@@ -1,4 +1,106 @@
...
+
+            mtime = os.path.getmtime(filename)
+
+            def uptodate():
+                try:
+                    return os.path.getmtime(filename) == mtime
+                except OSError:
+                    return False
+            return contents, filename, uptodate
+        raise TemplateNotFound(template)
```
Comment (kitsuta) [inline uber/jinja.py]: Just to double-check, will this class still work with the `ignore missing` keyword for the include tag? It's not a feature we use much (or at all...?) but it's a very nice feature.

File: uber/jinja.py (hunk 1 above)
Comment (RobRuana) [reply uber/jinja.py]: That exception would be passed up the stack, so we'd see the exception rendered. Most of the code in `MultiPathEnvironment` and `MultiPathLoader` was lifted wholesale from the Jinja source code

See: 
* https://github.com/pallets/jinja/blob/master/jinja2/loaders.py#L167
* https://github.com/pallets/jinja/blob/master/jinja2/environment.py#L795

File: uber/jinja.py (hunk 2 above)
Comment (RobRuana) [reply uber/jinja.py]: This should work the same way as the default jinja code - for the most part. Almost all the code in our custom classes is lifted straight from the jinja source.

**Comment by RobRuana** [comment]
I don't think this is over your head at all! You've authored some very tricky code yourself! If it's hard to follow, I should rewrite it to make it more clear.

The scariest part for me – the part that I was most surprised worked – is the code that determines which templates have already been used. We are examining the call stack and looking for files that end with ".html":
```python
    def _find_requesting_templates(self):
        """
        Returns the set of all files in the call stack ending in ".html".
        """
        stack = map(lambda s: s[0], traceback.extract_stack())
        return set(filter(lambda s: s.endswith('.html'), stack))
```

That may not be a great idea, but I'd love a second opinion...

**Comment by kitsuta** [comment]
I saw that! I wondered a little about it, but I don't have any better ideas... unless we can simply store templates in a Class property list as we load them?

**Comment by RobRuana** [comment]
Hmm... if we store them in a list in the class, how would we clear the list between requests? We want to store the templates used in just a single request, but I think the jinja `Environment` is re-used for many requests (the lifetime of the server, if I am correct).

**Comment by kitsuta** [comment]
Dang, I was kinda hoping the class would be re-initialized on requests or something, or that there'd at least be a hook we could use when a new request comes in. :(

**Comment by EliAndrewC** [comment]
I haven't had time to review this PR yet (I hope to do so tonight), but I'll note that Sideboard has a ``@request_cached_property`` (https://github.com/magfest/sideboard/blob/master/sideboard/lib/_utils.py#L87) which does exactly what @kitsuta wants and gives you a special property which re-initializes itself on every request.

I'll try to take a closer look at this PR later tonight.

**Comment by RobRuana** [comment]
I would have to look into the cherrypy internals before I'd feel confident that a thread local variable will never be reused by a subsequent request. Though, I know you've done a non-trivial amount of work on the cherrypy codebase.

I initially didn't write any unit tests for this bit of code mostly out of laziness. It wasn't immediately obvious to me how to test it properly. However, I've decided that I am not comfortable merging this *without* unit tests!

Hopefully I'll have some working tests before you get a chance to review this 😄

**Comment by EliAndrewC** [comment]
> I would have to look into the cherrypy internals before I'd feel confident that a thread local variable will never be reused by a subsequent request. Though, I know you've done a non-trivial amount of work on the cherrypy codebase.

They absolutely would be reused, except that Sideboard performs a reset at the beginning of each request specifically to allow us to do this kind of thing: 
https://github.com/magfest/sideboard/blob/master/sideboard/server.py#L18

You can see the implementation of the ``threadlocal`` utility at https://github.com/magfest/sideboard/blob/master/sideboard/websockets.py#L27

**Comment by RobRuana** [comment]
Interesting note: I tried implementing the same code using a `@request_cached_property`, only to receive `jinja2.exceptions.TemplateNotFound: macros.html`.

I turns out that `macros.html` is *NOT* added to the call stack when you do `{% import 'macros.html' as macros %}`. So we were able to load it multiple times when we were examining the call stack, because it wasn't listed in the call stack. This is cool, because it made the code work – but it's also bad, because it means `macros.html` was never cached.

I think this is ultimately the right approach – allowing each plugin to declare its own `registration.html` and extend the parent `registration.html`. However, I don't think either of the current implementations I've tried are exactly right (examining the call stack and using a `@request_cached_property`).

I am going to dive deeper into the jinja internals. At the point when `{% extends %}` is called jinja certainly knows the absolute path of the calling template. In order to differentiate two identically named templates from two plugins, all we need is the absolute path.

**Comment by RobRuana** [comment]
Let me amend my previous comment. It's not that `macros.html` is never cached, but that it's cached once for each template that imports it. So it will be loaded multiple times for each different request that uses `macros.html`, but then the cache will be used for subsequent requests.

**Comment by RobRuana** [comment]
OKAY! @EliAndrewC I didn't get to add any unit tests, because I was busy experimenting with different strategies.

However, as of [this commit](https://github.com/magfest/ubersystem/pull/2884/commits/b49475182032cf62ffe1857926747ca4df58ac14) I believe I have come up with a rock solid strategy. Feel free to review it whenever, and I'll add unit tests before merging.

**Comment by RobRuana** [comment]
OKAY! I added unit tests! Yay!

I am considering this pull request finished, however I would still like @EliAndrewC to review it before I merge it. The full pull request is only 248 lines, and the core changes are only about 80 lines. So it's actually small enough to be easily reviewable.

**Comment by EliAndrewC** [comment]
Indeed!  Sorry for the delay, I only left work about an hour ago and had a similarly long day yesterday, so I'm behind on my non-work stuff.  I'll try to get to this before bed, but should have no trouble reviewing it tomorrow at the latest.

File: uber/jinja.py [hunk 3]
```diff
@@ -1,4 +1,84 @@
...
+        return matching_filenames
+
+    def _load_template(self, name, globals):
+        """
+        Overridden to consider templates already loaded during the current request.
+        """
+        if self.loader is None:
+            raise TypeError('no loader for this environment specified')
+
+        loaded_templates = self._templates_loaded_for_current_request
```
Comment (EliAndrewC) [inline uber/jinja.py]: Random style nitpick: you declare ``loaded_templates`` here but then only use it once and later in this method you revert to just referencing ``self._templates_loaded_for_current_request`` directly.  You could probably just refer to ``loaded_templates`` throughout the function for both brevity and consistency.

File: uber/tests/templates/templates_2/test_extends_1.html [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{%- import 'test_import_macros.html' as macros -%}
+{%- extends 'test_extends_1.html' -%}
+{%- block main -%}
+{{ super() }}
```
Comment (EliAndrewC) [inline uber/tests/templates/templates_2/test_extends_1.html]: Sanity check: if I've understood things properly, then nothing we've done here would stop us from being able to call ``super()`` multiple times for multiple different ``block``s, right?  As far as I can tell, that should work perfectly fine, since it's the ``extends`` which causes the "parent" template to be loaded, and from there the ``super()`` call should just work the way it normally would without any problems.  I just wanted to make sure I had that right.

File: uber/jinja.py (hunk 3 above)
Comment (RobRuana) [reply uber/jinja.py]: Yeah, so I declared `loaded_templates` just to keep line-length down. But then later on I explicitly use `self._templates_loaded_for_current_request` to make clear that we are updating a persistent attribute of the object.

Here's my thinking, let me know how you feel about it. Even though `loaded_templates` and `self._templates_loaded_for_current_request` point to the same object in memory, if a new developer were reading the source code they would have to understand that context. So, it seemed like this:
```python
template = self.loader.load(self, filename, globals)
self._templates_loaded_for_current_request.add(template.filename)
```
was more obvious than this:
```python
template = self.loader.load(self, filename, globals)
loaded_templates.add(template.filename)
```

Because to truly understand what that line of code is doing, you must look up the declaration of `loaded_templates`. So maybe it would be better to not declare `loaded_templates` at all?

Perhaps I made the variable name for `_templates_loaded_for_current_request` way too long? I wanted to make it very obvious exactly what that property represented.

I don't know!

Comment (EliAndrewC) [reply uber/jinja.py]: That is indeed a tough "naming things" question.  I initially thought about suggesting that you rename ``self._templates_loaded_for_current_request`` to just be ``self._loaded_templates``, but then I decided against making that suggestion due to the fact that it might mislead someone into assuming that the templates were only loaded once period instead of per request.

I'm actually kind of a fan in some cases like this of having shorter and more cryptic names for private fields.  For example, if we named this ``self._rtc`` and then in the docstring said something like "rtc stands for 'request template cache', as this is the cache of templates which have been loaded for the currently-being-served request" then it would be short enough to keep the line length down without being misleading. 

One often-overlooked benefit of cryptic variable names like ``_rtc`` is that a newcomer seeing it for the first time won't read the variable name and immediately come to any misconceptions about what it is, because they *must* read the docstring to have any idea what it is!

This is obviously not a universally accepted or popular approach to naming, and it's not one that I would take for public fields on a class, but I'm personally a fan of doing it for private fields and methods.  Having longer variable names is more common (at my first job there was a variable named ``bulk_mail2_a0_width_increase_for_pass_trough_percent``), so it's mostly a style preference.
